project build/soong/
diff --git a/cc/config/arm_device.go b/cc/config/arm_device.go
index d37e4862..45a2183a 100644
--- a/cc/config/arm_device.go
+++ b/cc/config/arm_device.go
@@ -47,7 +47,7 @@ var (
 
 	armThumbCflags = []string{
 		"-mthumb",
-		"-Os",
+		"-O3",
 	}
 
 	armClangArchVariantCflags = map[string][]string{
diff --git a/cc/config/global.go b/cc/config/global.go
index 473c8067..381e56da 100644
--- a/cc/config/global.go
+++ b/cc/config/global.go
@@ -45,7 +45,11 @@ var (
 
 		"-O2",
 		"-g",
-
+             
+                "-fomit-frame-pointer",
+                "-frename-registers",
+                "-funswitch-loops",		
+		
 		"-fno-strict-aliasing",
 
 		"-Werror=date-time",
@@ -80,6 +84,7 @@ var (
 	}
 
 	deviceGlobalLdflags = []string{
+		"-Wl,--sort-common",
 		"-Wl,-z,noexecstack",
 		"-Wl,-z,relro",
 		"-Wl,-z,now",
diff --git a/cc/prebuilt.go b/cc/prebuilt.go
index b7c0bf2f..9c4602e4 100644
--- a/cc/prebuilt.go
+++ b/cc/prebuilt.go
@@ -27,7 +27,7 @@ func RegisterPrebuiltBuildComponents(ctx android.RegistrationContext) {
 	ctx.RegisterModuleType("cc_prebuilt_library_shared", PrebuiltSharedLibraryFactory)
 	ctx.RegisterModuleType("cc_prebuilt_library_static", PrebuiltStaticLibraryFactory)
 	ctx.RegisterModuleType("cc_prebuilt_object", prebuiltObjectFactory)
-	ctx.RegisterModuleType("cc_prebuilt_binary", prebuiltBinaryFactory)
+	ctx.RegisterModuleType("cc_prebuilt_binary", PrebuiltBinaryFactory)
 }
 
 type prebuiltLinkerInterface interface {
@@ -316,7 +316,7 @@ func (p *prebuiltBinaryLinker) binary() bool {
 
 // cc_prebuilt_binary installs a precompiled executable in srcs property in the
 // device's directory.
-func prebuiltBinaryFactory() android.Module {
+func PrebuiltBinaryFactory() android.Module {
 	module, _ := NewPrebuiltBinary(android.HostAndDeviceSupported)
 	return module.Init()
 }

